src/Functions/FunctionsPunycode.cpp

```diff
@@ -0,0 +1,329 @@
+#include <DataTypes/DataTypeString.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionStringToString.h>
+#include <IO/WriteBufferFromVector.h>
+#include <IO/WriteHelpers.h>
+#include "Poco/Unicode.h"
+#include <punycode.h>
+
+#include <string_view>
+#include <codecvt>
+#include <locale>
+#include <chrono>
+#include <thread>
+#include <cstring>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+}
+
+namespace
+{
+    // constexpr uint16_t domainMaxLength { 256 };
+    constexpr uint16_t labelMaxLength { 64 };
+    constexpr std::array<char, 4> acePrefix { 'x', 'n', '-', '-'};
+    std::wstring_convert<std::codecvt_utf8<char32_t>, char32_t> converter {};
+
+    template <uint16_t A, uint16_t B, class... Args>
+    constexpr bool inRange(const Args &... param) noexcept
+    {
+        return ((static_cast<char>(param) >= A && B >= static_cast<char>(param)) && ...);
+    }
+
+    template<typename... Args>
+    void insertToString(UInt8 * str, size_t & start, Args && ... args)
+    {
+        ((str[start++] = args), ...);
+    }
+
+    constexpr bool isASCII(const uint32_t u32) noexcept
+    {
+        if (u32 <= 0x7F) // 0XXXXXXX
+        {
+            return inRange<0, 127>(u32);
+        }
+        else if (u32 <= 0x7FF) // 110XXXXX 10XXXXXX
+        {
+            return inRange<0, 127>((0xC0 | ((u32 >> 6) & 0x1F)), (0x80 | (u32 & 0x3F)));
+        }
+        else if (u32 <= 0xFFFF) // 1110XXXX 10XXXXXX 10XXXXXX
+        {
+            return inRange<0, 127>((0xE0 | ((u32 >> 12) & 0x0F)), (0x80 | ((u32 >> 6) & 0x3F)),
+                                   (0x80 | (u32 & 0x3F)));
+        }
+        else if (u32 <= 0x13FFFF)  // 11110XXX 10XXXXXX 10XXXXXX 10XXXXXX
+        {
+            return inRange<0, 127>((0xF0 | ((u32 >> 18) & 0x07)), (0x80 | ((u32 >> 12) & 0x3F)),
+                                   (0x80 | ((u32 >> 6) & 0x3F)), (0x80 | (u32 & 0x3F)));
+        }
+        return false;
+    }
+
+    bool isASCIIStrUTF32(const std::u32string & input, size_t from, size_t until) noexcept
+    {
+        for (size_t i = from, size = std::min(input.size(), until); i < size; ++i)
+        {
+            if (!isASCII(input[i]))
+                return false;
+        }
+        return true;
+    }
+
+    bool isEncoded(const std::string_view input, size_t start) noexcept
+    {
+        return (input.length() >= (start + 4)) && (input[start] == 'x' ||input[start] == 'X') &&
+            (input[start + 1] == 'n' || input[start + 1] == 'N') &&
+            (input[start + 2] == '-' || input[start + 4] == '-');
```

Comment:
That doesn't seem completely correct. "xn--" is the prefix of IDNA encoding and not punycode. I think it's better to have 4 functions instead of 2:

```
punycodeEncode('ввв') -> 'b1aaa'
punycodeDecode()
idnaEncode('ввв.яндекс.рф') -> 'xn--b1aaa.xn--d1acpjx3f.xn--p1ai'
idnaDecode()
```